File: mkdocs/snippets/devbook/websockets/listen-new-blocks.py
import asyncio
import json
import os

from websockets import connect

NODE_URL = os.getenv('NODE_URL', 'https://reference.symboltest.net:3001')
WS_URL = NODE_URL.replace('http', 'ws', 1) + '/ws'
print(f'Using node {NODE_URL}')


async def main():
	async with connect(WS_URL) as websocket:  # [>step-1]
		# Connect to websocket endpoint
		response = json.loads(await websocket.recv())
		uid = response['uid']
		print(f'Connected to {WS_URL} with uid {uid}')
	# [<step-1]
		# Subscribe to block channel [>step-2]
		await websocket.send(json.dumps(
			{'uid': uid, 'subscribe': 'block'}))
		print('Subscribed to block channel')

		# Subscribe to finalizedBlock channel
		await websocket.send(json.dumps(
			{'uid': uid, 'subscribe': 'finalizedBlock'}))
		print('Subscribed to finalizedBlock channel')
		# [<step-2]
		# Handle incoming messages [>step-3]
		try:
			async for raw_message in websocket:
				message = json.loads(raw_message)
				topic = message['topic']

				if topic == 'block':
					block = message['data']['block']
					block_meta = message['data']['meta']
					print(
						f'New block: height={int(block["height"]):,}'
						f' hash={block_meta["hash"][:16]}...'
					)

				if topic == 'finalizedBlock':
					finalized = message['data']
					print(
						f'Finalized: height={int(finalized["height"]):,}'
						f' hash={finalized["hash"][:16]}...'
					)
		# [<step-3]
		# Unsubscribe on exit [>step-4]
		finally:
			await websocket.send(json.dumps(
				{'uid': uid, 'unsubscribe': 'block'}))
			await websocket.send(json.dumps(
				{'uid': uid, 'unsubscribe': 'finalizedBlock'}))
			print('Unsubscribed from all channels')

		# [<step-4]
try:
	asyncio.run(main())
except KeyboardInterrupt:
	pass
except Exception as error:
	print(error)

File: mkdocs/snippets/devbook/websockets/listen_new_blocks.mjs
const NODE_URL = process.env.NODE_URL
	|| 'https://reference.symboltest.net:3001';
const WS_URL = `${NODE_URL.replace('http', 'ws')}/ws`;
console.log(`Using node ${NODE_URL}`);

try {
	const websocket = new WebSocket(WS_URL); // [>step-1]

	// Connect to websocket endpoint
	const uid = await new Promise(resolve => {
		websocket.addEventListener('message', event => {
			const message = JSON.parse(event.data);
			resolve(message.uid);
		}, { once: true });
	});
	console.log(`Connected to ${WS_URL} with uid ${uid}`);
	// [<step-1]
	// Subscribe to block channel [>step-2]
	websocket.send(JSON.stringify({ uid, subscribe: 'block' }));
	console.log('Subscribed to block channel');

	// Subscribe to finalizedBlock channel
	websocket.send(JSON.stringify({ uid, subscribe: 'finalizedBlock' }));
	console.log('Subscribed to finalizedBlock channel');
	// [<step-2]
	// Handle incoming messages [>step-3]
	websocket.addEventListener('message', event => {
		const message = JSON.parse(event.data);
		const topic = message.topic;

		if ('block' === topic) {
			const block = message.data.block;
			const blockMeta = message.data.meta;
			console.log(
				'New block:'
				+ ` height=${parseInt(block.height, 10).toLocaleString()}`
				+ ` hash=${blockMeta.hash.substring(0, 16)}...`
			);
		}

		if ('finalizedBlock' === topic) {
			const finalized = message.data;
			console.log(
				'Finalized:'
				+ ' height='
				+ `${parseInt(finalized.height, 10).toLocaleString()}`
				+ ` hash=${finalized.hash.substring(0, 16)}...`
			);
		}
	});
	// [<step-3]
	// Unsubscribe on exit [>step-4]
	process.on('SIGINT', () => {
		websocket.send(JSON.stringify({ uid, unsubscribe: 'block' }));
		websocket.send(
			JSON.stringify({ uid, unsubscribe: 'finalizedBlock' }));
		console.log('Unsubscribed from all channels');
		websocket.close();
		process.exit(0);
	}); // [<step-4]
} catch (error) {
	console.error(error);
}
